pep-0xxx.rst

```diff
@@ -0,0 +1,797 @@
+            operation would have caused data to be written or read.
+
+            As at any time a re-negotiation is possible, a call to ``read()``
+            can also cause write operations.
+            """
+
+        @abstractmethod
+        def write(self, buf: Any) -> int:
+            """
+            Write ``buf`` in encrypted form to the output buffer and return the
+            number of bytes written. The ``buf`` argument must be an object
+            supporting the buffer interface.
+
+            Raise ``WantReadError`` or ``WantWriteError`` if there is
+            insufficient data in either the input or output buffer and the
+            operation would have caused data to be written or read.
+
+            As at any time a re-negotiation is possible, a call to ``write()``
+            can also cause read operations.
+            """
+
+        @abstractmethod
+        def do_handshake(self) -> None:
+            """
+            Performs the TLS handshake. Also performs certificate validation
+            and hostname verification.
+            """
+
+        @abstractmethod
+        def cipher(self) -> Optional[Cipher]:
+            """
+            Returns the Cipher entry for the cipher that has been negotiated on
+            the connection. If no connection has been negotiated, returns
+            ``None``.
+            """
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[NextProtocol]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @context.setter
+        @abstractmethod
+        def context(self, value: Context) -> None:
+            """
+            Set the value of the ``Context`` object this socket is tied to.
+            This operation (changing the context) may not always be supported.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+        @abstractmethod
+        def shutdown(self) -> None:
+            """
+            Performs a clean TLS shut down. This should generally be used
+            whenever possible to signal to the remote peer that the content is
+            finished.
+            """
+
+
+Cipher Suites
+~~~~~~~~~~~~~
+
+Todo
```

Comment:
Two ideas:

The OpenSSL cipher spec is not entirely unreasonable here, maybe we could make a few objects that basically implement that, but in a typed way. So that instead of:

```python
ciphers = "ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS"
```

You would get

```python
ciphers = (
    (
        ((Ciphers.ECDH + Ciphers.AESGCM) & (Ciphers.DH + Ciphers.AESGCM)) |
        ((Ciphers.ECDH + Ciphers.ChaCha20) & (Ciphers.DH + Ciphers.ChaCha20))
    ) &
    (Ciphers.ECDH + Ciphers.AES256) &
    (Ciphers.DH + Ciphers.AES256) &
    (Ciphers.ECDH + Ciphers.AES128) &
    (Ciphers.DH + Ciphers.AES) &
    (Ciphers.RSA + Ciphers.AESGCM) &
    (Ciphers.RSA + Ciphers.AES) &
    ~Ciphers.aNULL &
    ~Ciphers.eNULL &
    ~Ciphers.MD5 &
    ~Ciphers.DSS
)
```

The above has a quick sketch of the possibility to define "equivalent" ciphers through the use of grouping and the ``|`` statement. Obviously if the implementation doesn't support "equivalent" (do any of them? I know OpenSSL doesn't but I really wish it did...) they could just treat this as a flat list.

The other idea I had is using something like ``enum.Flag``.